### PR TITLE
add a network partition to Cardano HFC ThreadNet tests at interesting times

### LINKED ISSUE #2292: Test hard fork combinator near rollbacks
From https://github.com/input-output-hk/ouroboros-network/pull/2066#pullrequestreview-430336258

> Modify that tests (or set up a second one) so that we are making sure we are testing what happens near boundary bounds. Rollback across transition points, ledger forecast across transition points, that sort of thing. This will require some careful tweaking of the leadership schedule I think in order to get precisely the rollbacks that we expect. This might be a bit awkward to set up but it's important, because it's the only way we can really test what happens at those transition boundaries.

As of PR #2277, `ouroboros-consensus-cardano:test:test` tests a reasonably realistic HFC.

This Issue is to make that test more strenuous by ensuring rollback etc near the transition point.

### PR DESCRIPTION
**Pull request by nfrisby**
Closes #2292.

The individual commits are relatively self-contained.

`createConnectedChannelsWithDelay` and `chainSyncMiddle` are the main additions to `Network.hs` deserving of review. I'd appreciate any insights on how to refine/simplify those.

`ThreadNet/Cardano.hs` contains the bulk of the changes; lots of comments related to the tuning of the generators.

### CODE REVIEW

**Comment by nfrisby** [comment]
Here're the `tabulate`s from a run of 1000 tests cases.

```
    simple convergence:                       OK (3594.90s)
      +++ OK, passed 1000 tests.
      
      ReachesShelley label (1000 in total):
      61.5% pv Enabled, bs Enabled, ss Optional, True
      27.9% pv Enabled, bs Enabled, ss Required, True
       5.8% pv Blocked, bs Enabled, ss Optional, False
       3.3% pv Blocked, bs Enabled, ss Required, False
       1.4% pv Enabled, bs Blocked, ss Optional, False
       0.1% pv Blocked, bs Blocked, ss Optional, False
      
      average #txs/block (1000 in total):
      100.0% R_Btwn (0.0,1.0)
      
      final intersection depth, Byron, count (106 in total):
      98.1% 0 blocks
       0.9% 2 blocks
       0.9% 5 blocks
      
      final intersection depth, Byron, k diff (106 in total):
      34.0% k - 6 blocks
      19.8% k - 4 blocks
      17.9% k - 5 blocks
      17.0% k - 2 blocks
      10.4% k - 3 blocks
       0.9% k - 1 blocks
      
      final intersection depth, Byron, k frac (106 in total):
      98.1% k * 0.0 blocks
       0.9% k * 0.3 blocks
       0.9% k * 0.8 blocks
      
      final intersection depth, Shelley, count (894 in total):
      99.3% 0 blocks
       0.6% 1 blocks
       0.1% 2 blocks
      
      final intersection depth, Shelley, k diff (894 in total):
      21.0% k - 4 blocks
      20.1% k - 3 blocks
      19.9% k - 2 blocks
      19.5% k - 5 blocks
      19.2% k - 6 blocks
       0.2% k - 1 blocks
      
      final intersection depth, Shelley, k frac (894 in total):
      99.3% k * 0.0 blocks
       0.3% k * 0.2 blocks
       0.3% k * 0.5 blocks
      
      floor(4 * lastJoinSlot / numSlots) (1000 in total):
      100.0% Just 0
      
      involves >=1 re-delegation (1000 in total):
      100.0% False
      
      k (1000 in total):
      21.2% 6
      20.8% 4
      19.7% 2
      19.2% 5
      19.1% 3
      
      minimumDegreeNodeTopology (1000 in total):
      49.3% Just 1
      42.1% Just 2
       8.5% Just 3
       0.1% Just 4
      
      partition duration, count (1000 in total):
      18.5% 2 slots
      18.2% 0 slots
      17.5% 1 slots
      10.3% 3 slots
      10.3% 4 slots
       7.3% 5 slots
       7.0% 6 slots
       3.9% 8 slots
       3.6% 7 slots
       1.7% 10 slots
       1.7% 9 slots
      
      partition duration, k frac (1000 in total):
      18.2% k * 0.0 slots
      16.1% k * 1.0 slots
      13.1% k * 0.5 slots
       7.4% k * 0.2 slots
       7.4% k * 0.8 slots
       6.9% k * 1.2 slots
       6.5% k * 0.7 slots
       5.5% k * 0.3 slots
       5.5% k * 1.3 slots
       4.6% k * 1.5 slots
       2.4% k * 1.6 slots
       1.8% k * 0.4 slots
       1.7% k * 1.7 slots
       1.6% k * 1.4 slots
       1.3% k * 0.6 slots
      
      partition in or abuts era (Byron, Shelley) (1000 in total):
      37.4% (True,False)
      31.3% (True,True)
      18.2% (False,False)
      13.1% (False,True)
      
      shortestLength (k = 2) (197 in total):
      84.3% ≈ (k = 2) << 34
      15.7% ≈ (k = 2) << 66
      
      shortestLength (k = 3) (191 in total):
      59.7% ≈ (k = 3) << 67
      40.3% ≈ (k = 3) << 35
      
      shortestLength (k = 4) (208 in total):
      95.2% ≈ (k = 4) << 68
       4.8% ≈ (k = 4) << 132
      
      shortestLength (k = 5) (192 in total):
      91.1% ≈ (k = 5) << 69
       8.9% ≈ (k = 5) << 133
      
      shortestLength (k = 6) (212 in total):
      77.8% ≈ (k = 6) << 70
      22.2% ≈ (k = 6) << 134
```

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs [hunk 1]
```diff
@@ -392,13 +400,40 @@ prop_general ::
...
+     ( Condense blk
+     , Eq blk
+     , RunNode blk
+     )
+  => PropGeneralArgs blk
+  -> TestOutput blk
+  -> Property
+prop_general_semisync = prop_general_internal SemiSync
+
+data Synchronicity = SemiSync | Sync
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs:427]: A comment briefly explaining what you mean by synchrony and semi-synchrony would be helpful.

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs [hunk 2]
```diff
@@ -165,6 +171,22 @@ plainTestNodeInitialization pInfo = TestNodeInitialization
     , tniProtocolInfo = pInfo
     }
 
+-- | Compute the chain diffusion delay
+--
+-- This is the number of slots a 'CS.MsgRollForward' should arrive after the
+-- forge slot of the header it is carrying.
+--
+-- It may depend on the @(sender, recipient)@, the current slot, and header.
+newtype CalcMessageDelay blk = CalcMessageDelay
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs:181]: Something to consider for later, perhaps: do you envision doing any complicated calculation here, or would a mere tabulation (e.g. `[(CoreNodeId, CoreNodeId, SlotNo, NumSlots)]`) suffice? If so, that might be easier to `Show`, interpret in test failures, and, critically, shrink.

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs [hunk 3]
```diff
@@ -307,12 +331,18 @@ runThreadNetwork systemTime ThreadNetworkArgs
     -- fork the directed edges, which also allocates their status variables
     let uedges = edgesNodeTopology nodeTopology
     edgeStatusVars <- fmap (Map.fromList . concat) $ do
+      -- assume they all use the same CodecConfig
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs:335]: Just a heads-up, as of #2336 the `CodecConfig` is just another field in the top-level config (and we'll want to vary it from node to node once we start decontamination).

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs [hunk 4]
```diff
@@ -225,7 +225,6 @@ runTestNetwork TestConfig
           { nodeInfo
           , mkRekeyM
           } = mkTestConfigMB
-    setCurrentTime dawnOfTime
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs:224]: Not sure I understand this. How come this is redundant?

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs [hunk 5]
```diff
@@ -93,29 +110,165 @@ data TestSetup = TestSetup
...
+    -- of @firstSlotIn@ and ends at the onset of @firstSlotAfter@.
+
+    -- FACT (A) The leader of @firstSlotAfter@ will forge before fully reacting
+    -- to the @MsgRollForward@s that just arrived, due to a race in the test
+    -- infrastructure that is consistently won by the forge. Thus the two
+    -- class's unique extensions consist of the blocks forged in the slots from
+    -- @firstSlotIn@ to @firstSlotAfter@ /inclusive/.
+
+    -- how long the partition should last (this may be 'crop'ped if it's near
+    -- the end of the test)
```
Comment (edsko) [inline ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs:203]: Stray comment?

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs (hunk 4 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs:224]: `setCurrentTime dawnOfTime` also occurs five lines up in the same computation.

Comment (edsko) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs:224]: Ah :)

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs (hunk 2 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs:181]: I copied this over from the branch I've been developing semi-synchrony on for Praos. On that branch, this computation is driven by a random seed. That's where this shape originated. I'm hoping to develop that next and share this infra.

I don't have much intuition for the "size" of a set of delays. The main obstacle seems to be that the actual set of messages being delayed is unpredictable (especially for Praos). And so I'm punting on shrinking for now.

To your specific suggestion: `[(CoreNodeId, CoreNodeId, SlotNo, NumSlots)]` seems to assume that there is only one delay per directed edge per slot. But two nodes sometimes exchange multiple messages per slot. I'm unsure if it matters that those message potentially have different delays, but I'm currently favoring that they can so-as to make the fewest assumptions.

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/General.hs:427]: Good catch. Done! Thanks.

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs (hunk 5 above)
Comment (nfrisby) [reply ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs:203]: It's just the summary line at the start of a large comment. I've made that more obvious.

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs:335]: Ah, merged just in time 😛

**Comment by nfrisby** [comment]
bors merge

**Pull request merged**